### PR TITLE
Fix `test_melt` with `pyarrow` strings active

### PR DESCRIPTION
**Pull request by j-bennet**
Part of https://github.com/dask/dask/issues/10029.

Follow-up for https://github.com/dask/dask/pull/10000.

- [x] Tests added / passed
- [x] Passes `pre-commit run --all-files`

### CODE REVIEW

**Review by jrbourbeau — COMMENTED** [review]
Thanks @j-bennet

It's not totally clear to me whether this is a case where we should convert to `string[pyarrow]` or leave as `object` when combining mixed dtype columns (e.g. combining strings and floats). What's your thinking behind the current approach?

File: dask/dataframe/tests/test_multi.py [hunk 1]
```diff
@@ -216,6 +216,17 @@ def list_eq(aa, bb):
...
+
+    if dask.config.get("dataframe.convert_string"):
+        from dask.dataframe._pyarrow import to_pyarrow_string
+
+        if isinstance(a, (pd.DataFrame, pd.Series, pd.Index)):
+            a = to_pyarrow_string(a)
+
+        if isinstance(b, (pd.DataFrame, pd.Series, pd.Index)):
+            b = to_pyarrow_string(b)
+
```
[the remark this answers is not in the document]
Comment (j-bennet) [reply dask/dataframe/tests/test_multi.py:229]: 👍

**Comment by j-bennet** [comment]
> It's not totally clear to me whether this is a case where we should convert to `string[pyarrow]` or leave as `object` when combining mixed dtype columns (e.g. combining strings and floats). What's your thinking behind the current approach?

For some reason, I had an impression that `pandas` would lose `string[pyarrow]` dtypes when melting them together, but this turns out to not be the case (perhaps I've seen this in the past, but it was fixed? not sure). Pandas has reasonable upcasting behavior:

```
obj + str = obj
obj + int = obj
str + str = str
str + int = obj
```

so the best fix seems to be to turn off `convert_string` and delegate to upstream. This PR is now doing just that.

File: dask/dataframe/utils.py [hunk 2]
```diff
@@ -544,6 +543,14 @@ def assert_eq_dtypes(a, b):
...
             b = to_pyarrow_string(b)
 
+    return a, b
+
+
+def assert_eq_dtypes(a, b):
+    (
+        a,
+        b,
+    ) = _maybe_convert_string(a, b)
```
Comment (jrbourbeau) [inline dask/dataframe/utils.py:553]: I've not checked locally -- is `pre-commit` happy with this formatting? Not a big deal if we need what's current in this PR, but just thought I'd check

```suggestion
    a, b = _maybe_convert_string(a, b)
```

File: dask/dataframe/tests/test_multi.py [hunk 3]
```diff
@@ -1627,38 +1627,35 @@ def fix_index(out, dtype):
...
+        {
+            "obj": list("abcd") * 5,
+            "s1": list("XY") * 10,
+            "s2": list("abcde") * 4,
+            "int": np.random.randn(20),
+        }
     )
-    ddf = dd.from_pandas(pdf, 4)
+    pdf.s1 = pdf.s1.astype(pd.StringDtype("pyarrow"))
+    pdf.s2 = pdf.s2.astype(pd.StringDtype("pyarrow"))
```
Comment (jrbourbeau) [inline dask/dataframe/tests/test_multi.py:1652]: Is this intended, or left over from local development?

File: dask/dataframe/reshape.py [hunk 4]
```diff
@@ -358,13 +359,15 @@ def melt(
...
-        meta=no_default,
-        id_vars=id_vars,
-        value_vars=value_vars,
-        var_name=var_name,
-        value_name=value_name,
-        col_level=col_level,
-        token="melt",
-    )
+    # let pandas do upcasting as needed during melt
+    with config.set({"dataframe.convert_string": False}):
```
Comment (jrbourbeau) [inline dask/dataframe/reshape.py:363]: Not a big deal, or meant as a blocking comment, but I normally see 

```python
import dask
...
dask.config.get(...)
dask.config.set(...)
```

File: dask/dataframe/utils.py (hunk 2 above)
Comment (j-bennet) [reply dask/dataframe/utils.py:553]: This is what pre-commit did, for whatever reason! I didn't do that. :)

File: dask/dataframe/tests/test_multi.py (hunk 3 above)
Comment (j-bennet) [reply dask/dataframe/tests/test_multi.py:1652]: This is intended, I want `pandas` to melt together some `pyarrow` strings.

File: dask/dataframe/reshape.py (hunk 4 above)
Comment (j-bennet) [reply dask/dataframe/reshape.py]: Sure, that's an easy change.

File: dask/dataframe/tests/test_multi.py (hunk 3 above)
Comment (jrbourbeau) [reply dask/dataframe/tests/test_multi.py:1652]: Ah, I missed that these are new columns you're adding. In that case, this looks fine. I would add some additional logic to handle the case that `pyarrow` isn't installed and maybe simplify to 

```python
pdf = pdf.astype({"s1": "string[pyarrow]", "s2": "string[pyarrow]"})
```

File: dask/dataframe/utils.py (hunk 2 above)
Comment (jrbourbeau) [reply dask/dataframe/utils.py:553]: Hrm, weird. When I make this change, `pre-commit` seems to be happy locally

Comment (j-bennet) [reply dask/dataframe/utils.py]: Yeah, it's happy now.

File: dask/dataframe/tests/test_multi.py (hunk 3 above)
Comment (jrbourbeau) [reply dask/dataframe/tests/test_multi.py]: I went ahead and pushed a tiny commit that makes sure `test_melt` is run, even when `pyarrow` isn't installed. Hope that's okay

**Review by jrbourbeau — APPROVED** [review]
Thanks @j-bennet -- will merge after CI finishes

**Comment by j-bennet** [comment]
Test failure is unrelated (upstream): `FAILED dask/dataframe/tests/test_pyarrow_compat.py::test_pickle_roundtrip[uint8]`.

**Pull request merged**